### DISCUSSION TITLE
Any plans to submit changes upstream?

### DISCUSSION THREAD

**Discussion by trallen** [jakobheyman/xdripJH#6] [General]
Hi @jakobheyman , do you have any plans to submit any of your diffs upstream for inclusion into xDrip directly? Is there anything I could do to help?

Thanks!

**Comment by jakobheyman** [comment]
Thanks for your question!

I have thought sometimes that I should at least write in the official xDrip discussions and offer that I could try to make pull requests for some of my changes, but I have not done it yet. I would guess that many of the changes that I have done are too radical for the official xDrip, but there are some changes that I see as harmless and that I imagine could be of value to several xDrip users. Apart from the Libre-2 one minute data I think there are three main changes that could be useful for many:
1. That the Y range is kept constant in the main graph.
2. The option to set custom vibration patterns for alarms.
3. Custom start and stop dates for the statistics.

For point 1, I did mention it in some discussion topic in main xDrip but there was not much enthusiasm about it so I did not take it further. For points 2 and 3, my changes of the code are probably rather ugly as they are just made by copying other parts and trying to just make it work. Therefore, I have not mentioned them upstream as at least presently I would not have time enough to try to fix the code better. If there would be a wish from someone that I should submit some change upstream I can try. You are also most welcome to take whichever parts you like and try to fix the code nicely and submit to main xDrip.

**Reply by trallen** [reply]
> I have thought sometimes that I should at least write in the official xDrip discussions and offer that I could try to make pull requests for some of my changes, but I have not done it yet. I would guess that many of the changes that I have done are too radical for the official xDrip, but there are some changes that I see as harmless and that I imagine could be of value to several xDrip users. Apart from the Libre-2 one minute data I think there are three main changes that could be useful for many:

I'd be delighted to see the one minute data changes in core xDrip. I know jamorham is working on it in https://github.com/NightscoutFoundation/xDrip/issues/2680 but it doesn't seem to be there yet. I'm slowly reviewing and adopting your changes into my local branch over that code, but I'm not fully there either!
 
>     1. That the Y range is kept constant in the main graph.

I see that the code in upstream has this commit: https://github.com/NightscoutFoundation/xDrip/commit/749f451ea6c0adfadf502891f5e62a6bb53b6254 ... maybe yours could be integrated with those changes?

(I have been working on the X range, in https://github.com/NightscoutFoundation/xDrip/pull/3464 .)

> For point 1, I did mention it in some discussion topic in main xDrip but there was not much enthusiasm about it so I did not take it further. For points 2 and 3, my changes of the code are probably rather ugly as they are just made by copying other parts and trying to just make it work. Therefore, I have not mentioned them upstream as at least presently I would not have time enough to try to fix the code better. If there would be a wish from someone that I should submit some change upstream I can try. You are also most welcome to take whichever parts you like and try to fix the code nicely and submit to main xDrip.

Thanks! I will give it a go; my main concern is that I might not be able to explain the reasoning behind certain sections; if you were amenable to following any eventual pull request that I make and offering clarification where needed, I'd be most grateful.

But fundamentally, I don't want to upstream your code without crediting you!

**Reply by jakobheyman** [reply]
Thanks!
> I'd be delighted to see the one minute data changes in core xDrip.

The one minute data was what I started with in January 2021 and I think many changes have been done in small steps over time. When I started I was using the patched LibreLink app but I then changed to using OOP2 (I believe in June 2021). At some point I also modified the calibration quite heavily and I guess that some changes related to the one minute data could be in changes that I have done to the calibration or so. Hence, the one minute data changes are perhaps a bit tricky to pick out from the github history. Anyway, I would be happy to try to help fixing so that one minute data could be added to core xDrip. I seem to remember that there is at least one issue with my changes that I have not fixed probably (from some short comments in gitter). I believe that it was something with predictions (but it could also be something else - it was something that I don't use myself) that does not work properly in xdripJH as I have not fixed the code there for 1 or 5 min data properly.

I will write more later.

**Comment by jakobheyman** [comment]
> I see that the code in upstream has this commit: https://github.com/NightscoutFoundation/xDrip/commit/749f451ea6c0adfadf502891f5e62a6bb53b6254 ... maybe yours could be integrated with those changes?

Yes - maybe. The changes to the viewport that I have done is one of the changes that I am most proud of and a change that one can notice as it directly impacts the view of the glucose curve. I think the main parts that I have changed (although there are maybe some more) can be found here:
[Home.java](https://github.com/jakobheyman/xdripJH/blob/5d219059cb9ce8df37ee9db8b7fe06cc5ba13cf3/app/src/main/java/com/eveningoutpost/dexdrip/Home.java#L2299C1-L2330C101)
[BgGraphBuilder.java](https://github.com/jakobheyman/xdripJH/blob/5d219059cb9ce8df37ee9db8b7fe06cc5ba13cf3/app/src/main/java/com/eveningoutpost/dexdrip/utilitymodels/BgGraphBuilder.java#L2206C1-L2255C10)

> Thanks! I will give it a go; my main concern is that I might not be able to explain the reasoning behind certain sections; if you were amenable to following any eventual pull request that I make and offering clarification where needed, I'd be most grateful.

Great! I will do my best to try to follow any future pull requests and help if needed.

> But fundamentally, I don't want to upstream your code without crediting you!

Thanks!
However, I don't think that I would care much if the code that I have written would be used without me getting credit for it. I am just very happy that xDrip exists as open source software that I can modify and track with git.
